src/main/kotlin/org/rust/lang/core/macros/MacroExpansionVfsBatch.kt

```diff
@@ -41,154 +37,71 @@ class MacroExpansionVfsBatchImpl(rootDirName: String) : MacroExpansionVfsBatch {
+    fun createFile(parent: String, name: String, content: String): String {
         val child = "$parent/$name"
-        MacroExpansionFileSystem.getInstance().createDirectory(child)
-        dirEvents.add(DirCreateEvent(parent, name))
+        MacroExpansionFileSystem.getInstance().createFileWithContent(child, content, mkdirs = true)
+        pathsToMarkDirty += parent
         return child
     }
 
-    fun String.getOrCreateDirectory(name: String): String =
-        getOrCreateDirectory(this, name)
-
-    @JvmName("getOrCreateDirectory_")
-    private fun getOrCreateDirectory(parent: String, name: String): String {
-        val child = "$parent/$name"
-        if (MacroExpansionFileSystem.getInstance().exists(child)) {
-            return child
-        }
-        return createDirectory(parent, name)
+    fun writeFile(file: VirtualFile, content: String) {
+        MacroExpansionFileSystem.getInstance().setFileContent(file.path, content)
+        markDirty(file)
     }
 
-    fun writeFile(file: String, content: String) {
-        MacroExpansionFileSystem.getInstance().setFileContent(file, content)
-
-        fileEvents.add(Event.Write(file))
+    fun deleteFile(file: VirtualFile) {
+        MacroExpansionFileSystem.getInstance().deleteFile(file.path)
+        markDirty(file)
     }
 
-    fun deleteFile(file: String) {
-        MacroExpansionFileSystem.getInstance().deleteFile(file)
+    fun applyToVfs(async: Boolean, callback: Runnable? = null) {
+        val root = MacroExpansionFileSystem.getInstance().findFileByPath("/") ?: return
 
-        fileEvents.add(Event.Delete(file))
-    }
-
-    abstract fun applyToVfs()
-
-    protected class DirCreateEvent(val parent: String, val name: String)
-
-    protected sealed class Event {
-        class Create(val parent: String, val name: String): Event()
-        class Write(val file: String): Event()
-        class Delete(val file: String): Event()
-    }
-
-}
-
-class EventBasedVfsBatch : VfsBatch() {
-    override fun applyToVfs() {
-        checkWriteAccessAllowed()
-        if (dirEvents.isEmpty() && fileEvents.isEmpty()) return
-
-        val manager = VirtualFileManager.getInstance() as VirtualFileManagerEx
-        manager.fireBeforeRefreshStart(/*asynchronous = */ false)
-        try {
-            val events = mutableListOf<VFileEvent>()
-            while (dirEvents.isNotEmpty()) {
-                val iter = dirEvents.iterator()
-                while (iter.hasNext()) {
-                    val event = iter.next().toVFileEvent()
-                    if (event != null) {
-                        iter.remove()
-                        events.add(event)
-                    }
-                }
-                check(events.isNotEmpty())
-                PersistentFS.getInstance().processEvents(events)
-                events.clear()
+        for (path in pathsToMarkDirty) {
+            var file = root
+            for (segment in StringUtil.split(path, "/")) {
```

Comment:
```suggestion
            for (segment in path.split("/")) {
```

Reply:
Hmm, looks like `path.split('/')` and `StringUtil.split(path, "/")` works differently: `path.split('/')` produces empty strings.

Reply:
Oh, it's interesting difference
Ok, let's revert the corresponding changes then